File: testdata/fixtures/php/CWE-601b.php
<?php
// CWE-601: Open Redirect via header Location with $_GET
function vuln1() {
    $url = $_GET['next'];
    // ruleid: php-open-redirect
    header('Location: ' . $url);
}

function vuln2() {
    // ruleid: php-open-redirect
    header('Location: ' . $_GET['url']);
}

function vuln3() {
    $target = $_POST['target'];
    // ruleid: php-open-redirect
    header('Location: ' . $target);
}

function safeLiteral() {
    // ok: php-open-redirect
    header('Location: /home');
}

function safeAllowlist() {
    $url = $_GET['next'];
    if ($url == '/home' || $url == '/dashboard') {
        // ok: php-open-redirect
        header('Location: /home');
    }
}
?>

File: testdata/fixtures/kotlin/CWE-601b.kt
// CWE-601b: Open redirect via request.getParameter("next") without is_safe_url
package fixture

import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class RedirectHandler {
    fun vuln1(request: HttpServletRequest, response: HttpServletResponse) {
        val next = request.getParameter("next")
        // ruleid: kotlin-601-fix
        response.sendRedirect(next)
    }

    fun vuln2(request: HttpServletRequest, response: HttpServletResponse) {
        val url = request.getParameter("url")
        // ruleid: kotlin-601-fix
        redirect(url)
    }

    fun vuln3(request: HttpServletRequest, response: HttpServletResponse) {
        val dest = request.getParameter("next")
        // ruleid: kotlin-601-fix
        call.respondRedirect(dest)
    }

    fun safeAllowlist(request: HttpServletRequest, response: HttpServletResponse) {
        val next = request.getParameter("next")
        if (!is_safe_url(next)) return
        // ok: kotlin-601-fix
        response.sendRedirect("/home")
    }

    fun safeFixed(request: HttpServletRequest, response: HttpServletResponse) {
        // ok: kotlin-601-fix
        response.sendRedirect("/home")
    }

    fun is_safe_url(u: String): Boolean { return u.startsWith("/") }
    fun redirect(u: String) {}
    val call = object { fun respondRedirect(u: String) {} }
}
